tasks/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+                            trickle_down(body, self.active_tasks_mut(), |task, body| {
+                                task.process_untagged(body)
+                            })
+                        {
+                            return SchedulerEvent::Unsolicited(Response::Status(Status::from(
+                                StatusV2::Untagged(body),
+                            )));
+                        }
+                    }
+                    StatusV2::Bye(bye) => {
+                        if let Some(bye) =
+                            trickle_down(bye, self.active_tasks_mut(), |task, bye| {
+                                task.process_bye(bye)
+                            })
+                        {
+                            return SchedulerEvent::Unsolicited(Response::Status(Status::from(
+                                StatusV2::Bye(bye),
+                            )));
+                        }
+                    }
+                    StatusV2::Tagged(Tagged { tag, body }) => {
+                        let handle = {
+                            let (handle, _) = self
+                                .active_tasks
+                                .iter()
+                                .find(|(_, (tag_, _))| tag == *tag_)
+                                .unwrap();
+
+                            *handle
+                        };
+
+                        let (_, task) = self.active_tasks.remove(&handle).unwrap();
+
+                        let output = Some(task.process_tagged(body));
+
+                        return SchedulerEvent::TaskFinished(TaskToken { handle, output });
+                    }
+                },
+            }
+        }
+    }
+
+    fn active_tasks_mut(&mut self) -> impl Iterator<Item = &mut Box<dyn TaskAny>> {
+        self.active_tasks.values_mut().map(|(_, task)| task)
+    }
+}
+
+#[derive(Debug)]
+pub enum SchedulerEvent {
+    TaskFinished(TaskToken),
+    Shutdown,
+    Unsolicited(Response<'static>),
+}
+
+#[derive(Eq)]
+pub struct TaskHandle<T: Task> {
+    handle: ClientFlowCommandHandle,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: Task> Debug for TaskHandle<T> {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        f.debug_struct("TaskHandle")
+            .field("handle", &self.handle)
+            .finish()
+    }
+}
+
+impl<T: Task> Clone for TaskHandle<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<T: Task> Copy for TaskHandle<T> {}
+
+impl<T: Task> PartialEq for TaskHandle<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.handle == other.handle
+    }
+}
+
+impl<T: Task> TaskHandle<T>
+where
+    T: 'static,
+{
+    fn new(handle: ClientFlowCommandHandle) -> Self {
+        Self {
+            handle,
+            _phantom: Default::default(),
+        }
+    }
+
+    /// Try resolving the task invalidating the token.
+    ///
+    /// The token is invalidated iff the return value is `Some()`.
```

Comment:
I prefer `Some` instead of this pseudo-syntax.

Reply:
Btw., did you know you can use a "suggestion" block in GitHub like ...

```suggestion
    /// The token is invalidated iff the return value is `Some`.
```

It's a bit faster than to explain simple changes, and a PR author can automatically accept all suggestions in a batch.

Reply:
Yeah, I'm just lazy :p